core/src/main/scala/cats/Apply.scala

```diff
@@ -257,6 +257,30 @@ trait Apply[F[_]] extends Functor[F] with InvariantSemigroupal[F] with ApplyArit
     def ite(b: Boolean)(ifTrue: A, ifFalse: A) = if (b) ifTrue else ifFalse
     ap2(map(fcond)(ite))(ifTrue, ifFalse)
   }
+
+  def select[A, B](fab: F[Either[A, B]])(ff: => F[A => B]): F[B] =
+    selectA(fab)(ff)
+
+  def selectA[A, B](fab: F[Either[A, B]])(ff: F[A => B]): F[B] =
```

Comment:
Remind me again why we need both select and selectA if setting them equal is lawful? Why not just add select which may or may not be override to be rigid?

Reply:
* `selectA` is like `ifA`: requires 'Apply`, not rigid even for `FlatMap`s
* `selectM` (not added yet) is like `ifM`: requires `FlatMap`, rigid
* `select` is like `ifS`: requires 'Apply`, rigid for `FlatMap`s, otherwise effect's choice

I can live without `selectA` and `selectM`, but their existence parallels some other functions.  (I'll note that @SystemFw just dumped on `ifA` on Gitter today.)

Reply:
I suspect ifA without rigidity is a recipe for exponential blow ups in cost.

Reply:
>  (I'll note that @SystemFw just dumped on ifA on Gitter today.

I have barely skimmed the whole conversation, so I do apologise for the lack of context, but I do think `ifA` is one of the worse things added in cats.
An `if` that executes both branches is what you want 0.01% of the time, and users have been told to "use the least powerful constraint you can", which makes it a recipe for disaster. We can't even claim we didn't see this, because the `if` + macro interaction in sbt has been hated for the same reason for years.

Personally I see Selective as just pulling things down from Monad, to get you more static analysis, so in an ideal world `ifA` would be gone (and similarly for `select*` on Apply), `Selective` would called just that, and have `ifM` (`ifS`?), etc. So (barring misunderstanding of what I've skimmed), I think I'm largely in agreement with @johnynek

Reply:
So, I caused a bit of the wild goose chase here I think.

One thing I was and am concerned about is a lawful implementation of Selective using Applicative, why not add to Applicative? So, I pushed to lower things to Applicative or Apply. But that means the default implementation winds up executing both sides, but as you note, you would almost never want that. It seems weird to open the door for exactly one use case that we can name (over approximating dependencies in a modestly dynamic graph).

So, now my thinking has gotten to either:

1. Reintroduce Selective (basically back where we started, yes you could implement Selective with Applicative, but it is probably not what you want, so it isn't a default, this allows you to implement it when you are sure that is what you want).
2. I'm dubious RigidSelective is much of a thing. I tend to wonder if the only instances are things that could be monads but are being restricted in some way (e.g. CLI parsers being restricted so that they can generate a static help).
3. A conservative way out would be to add `def select` to Monad and implement it with flatMap, add the select laws there (and require it to be rigid). This allows Monad instances to optimize select if they can (like Parser or Gen could).

So, one idea would be:

```scala
// Comment that this is a rarely used type that is generally used to make static analysis of
// F[_] values that model a deferred computation.
trait Selective[F[_]] extends Applicative[F] {
  def select[A, B](e: F[Either[A, B]])(fn: F[A => B]): F[B]
}

trait RigidSelective[F[_]] extends Selective[F] {
  def ifM... // all the monad methods that can be implemented in terms of select
}

trait Monad[F[_]] extends RigidSelective[F] {
...
}
```

Reply:
I hadn't seen Oscar's reply while doing this research, but dumping it while the kids are entertained.

The `select == selectM` relationship for monads has never been controversial.  And I think `ap == apS` follows for flatMaps.  Every `Apply` can be selective (associativity) with more laws (identity, distributivity) once `Applicative`.

So we should focus on instances that are `Apply` but not `FlatMap`.  There are 55 such implementations in cats and cats-effect.  Several are inductive instances.

What's left are worth thinking about, because they each:
1. May have an interesting `select != selectA`.
2. May have an offensive `select` that justifies a `Selective` typeclass.

## The data types

### Parallel versions of otherwise monadic types

We already identified `Validated` as an interesting `Selective`.  I bet all of these are interesting, too:

* cats.data.NonEmptyList.ZipNonEmptyList
* cats.data.NonEmptySeq.ZipNonEmptySeq
* cats.data.NonEmptyVector.ZipNonEmptyVector
* cats.data.Validated
* cats.data.ZipLazyList
* cats.data.ZipList
* cats.data.ZipSeq
* cats.data.ZipStream
* cats.data.ZipVector
* cats.effect.IO.Par

### Static analysis

* cats.data.Const: has two interesting selectives, and the only known interesting one where `select == selectA`.

### Data types we need to think harder about

* cats.data.AppFunc
* cats.data.Func
* cats.data.Nested 
* cats.effect.Fiber

## Type classes to think about

* cats.arrow.Arrow: we (non-implicitly) derive Applicative from Arrow. What does this imply about its default `Selective`?  Can we derive a more interesting one?

## Composed instances

Could there be a `ComposedSelective` to match? This is closely related to `Nested`.

* cats.ComposedApply
* cats.ComposedApplicative
* cats.ComposedAlternative

Reply:
Great summary Ross.

To simplify, I think ZipList, NonEmptyList.ZipNonEmptyList, Validated, IO.Par and Nested are the core things to think about. Clearly if Zip*List works, the other collections which are isomorphic should work too.

Reply:
I need to cook now, but quick response to "So, one idea would be:..."

* Yes, let's set aside the difference between `Apply` and `Applicative` for now.  We can circle back on maps and tuples in a second PR, or not, after we're done learning from this.
* Agree with reviving `Selective[F]`.  Several of the laws (identity, associativity, distributivity) can stick there.
* Agree on monad having rigidity laws.  (That has remained constant through this PR.)

I have some qualms pulling the `*M` functions down to `RigidSelective`:
1. As you note, few things are without being monads.  Which things?  Depends whether we stick to our skip laws, or accept the `ap == apS` law, or both.
2. Depending how we resolve the above, some things may have a viable `ifM` without being a `RigidSelective`, but we'd have no typeclass for these.  We could pull those functions down to `Selective`, but the name would lose its current sense of rigidity.
3. Some of the`'*M` functions are on `FlatMap`.  Now we need to think about that relationship earlier than we'd like.

Haskell gets around all these problems with an `*S` set of functions that's usually like `*M`. Great for laws, confusing for new users autocompleting weird names.

Reply:
Oof. `ZipList` is in the pile of things that are `Apply` but not `Applicative` that I'm trying to forget about for now.

`ZipStream`, `ZipLazyList`, and `IO.Par` are all `Applicative`.

Reply:
The Applicative[ZipLazyList] laws don't terminate, so I don't know if our `Selective[ZipLazyList]` is lawful.  But it appears to be Haskell-rigid: the `Apply` laws pass when `ap` is implemented as `apS`.  It's not Cats-Issue-3709-rigid, because we have to evaluate `ff` even if we get a lazy stream of Rights.

Reply:
> 2. I'm dubious RigidSelective is much of a thing. I tend to wonder if the only instances are things that could be monads but are being restricted in some way (e.g. CLI parsers being restricted so that they can generate a static help).

One very good reason for having a specific `Selective` thing that _isn't_ also a monad is precisely to leverage static structure with dynamic behaviours. Take parsers, for example: `Applicative` isn't strong enough to do context-sensitive things, `Monad` is, and `Selective` gives you some (of the most common) context-sensitive behaviours but not all of them. This is important, because if you wanted to generate fast hand-written recursive descent for a parser combinator implementation, you need a fully static structure. That rules out `Monad`, but not `Selective` or `Applicative`: luckily these can almost always be enough for your practical parsing needs, but such a library wouldn't be able to make a `Monad` instance (see https://github.com/j-mie6/ParsleyHaskell as an example of such a library)